BUG: sorting with tuples and Multi-Index creation

Hello,

I'm using python 2.7.9 and pandas 0.15.2.
I'm trying to concatenate two DataFrames (to_concat is a list of 2 DataFrames) and I have the following error.
```
  File "/Users/alexisbenoist/Documents/CoTH/AFDS/app/core.py", line 67, in train_dataflow
    df = pd.concat(to_concat)
  File "/Users/alexisbenoist/.virtualenvs/afds/lib/python2.7/site-packages/pandas/tools/merge.py", line 725, in concat
    return op.get_result()
  File "/Users/alexisbenoist/.virtualenvs/afds/lib/python2.7/site-packages/pandas/tools/merge.py", line 896, in get_result
    mgrs_indexers, self.new_axes, concat_axis=self.axis, copy=self.copy)
  File "/Users/alexisbenoist/.virtualenvs/afds/lib/python2.7/site-packages/pandas/core/internals.py", line 4046, in concatenate_block_managers
    for placement, join_units in concat_plan]
  File "/Users/alexisbenoist/.virtualenvs/afds/lib/python2.7/site-packages/pandas/core/internals.py", line 4135, in concatenate_join_units
    empty_dtype, upcasted_na = get_empty_dtype_and_na(join_units)
  File "/Users/alexisbenoist/.virtualenvs/afds/lib/python2.7/site-packages/pandas/core/internals.py", line 4124, in get_empty_dtype_and_na
    raise AssertionError("invalid dtype determination in get_concat_dtype")
AssertionError: invalid dtype determination in get_concat_dtype
```

I put the pickle of the DataFrames in a [gist](https://gist.github.com/Alexis-benoist/ba7f38deccc70d46c32f) using (I can upload it somewhere else if it's more convenient).
 
```
import pickle
with open(path, 'w+') as f:
    pickle.dump(to_concat, f)
```

I feel that I'm doing something wrong or it's a bug from pandas because if do: 
```
pd.DataFrame(np.concatenate(to_concat), columns=to_concat[0].columns.tolist())
```
It gives a result which seems logical.

Any input is welcome.

Cheers,
Alexis.

This was a bit tricky to repro. However, the bottom line is their is a small edge case bug that you hit, though you should NOT ever use tuples in an index, unless it really is a multi-index, as these are by-definition the markers for multi-index guessing.

This is very odd in how things are sorted, not sure why exactly python does this (this is py2).
```
In [23]: sorted([u'class',(u'internal',u'__delete__')])
Out[23]: [(u'internal', u'__delete__'), u'class']

In [24]: sorted(['class',('internal','__delete__')])
Out[24]: ['class', ('internal', '__delete__')]
```

So this makes a determination if the guessed index is a multi-index or not (this is the bug part, I think we can be more strict, IOW we are only checking the first element if its a tuple; if it is we try to create a multi-index).
```
In [29]: Index([('A',1),'B'])
In [29]: Index([('A',1),'B'])
Out[29]: 
MultiIndex(levels=[[u'A', u'B'], [1]],
           labels=[[0, 1], [0, -1]])

In [30]: Index(['B',('A',1)])
Out[30]: Index([u'B', (u'A', 1)], dtype='object')
```
Here's a repro example. Note that if these are strings (e.g. no ``u``), then this would WORK!.
```
In [26]: df1 = DataFrame({ u'A' : 'foo', (u'B',1) : 'bar' },index=range(2))

In [27]: df2 = DataFrame({ u'B' : 'foo', (u'B',1) : 'bar' },index=range(2))

In [28]: pd.concat([df1,df2])
AssertionError: invalid dtype determination in get_concat_dtype
```

--- a/doc/source/whatsnew/v0.17.0.txt
+++ b/doc/source/whatsnew/v0.17.0.txt
@@ -367,7 +367,7 @@ Bug Fixes
 - Bug in ``Index.drop_duplicates`` dropping name(s) (:issue:`10115`)
 - Bug in ``pd.Series`` when setting a value on an empty ``Series`` whose index has a frequency. (:issue:`10193`)
 - Bug in ``DataFrame.plot`` raises ``ValueError`` when color name is specified by multiple characters (:issue:`10387`)
-
+- Bug in ``Index`` construction with a mixed list of tuples (:issue:`10697`)
 - Bug in ``DataFrame.reset_index`` when index contains `NaT`. (:issue:`10388`)
 - Bug in ``ExcelReader`` when worksheet is empty (:issue:`6403`)
 - Bug in ``Table.select_column`` where name is not preserved (:issue:`10392`)

--- a/pandas/core/index.py
+++ b/pandas/core/index.py
@@ -164,16 +164,21 @@ class Index(IndexOpsMixin, PandasObject):
             cls._scalar_data_error(data)
         else:
             if tupleize_cols and isinstance(data, list) and data and isinstance(data[0], tuple):
-                try:
 
-                    # must be orderable in py3
-                    if compat.PY3:
-                        sorted(data)
-                    return MultiIndex.from_tuples(
-                        data, names=name or kwargs.get('names'))
-                except (TypeError, KeyError):
-                    # python2 - MultiIndex fails on mixed types
-                    pass
+                # we must be all tuples, otherwise don't construct
+                # 10697
+                if all( isinstance(e, tuple) for e in data ):
+
+                    try:
+
+                        # must be orderable in py3
+                        if compat.PY3:
+                            sorted(data)
+                        return MultiIndex.from_tuples(
+                            data, names=name or kwargs.get('names'))
+                    except (TypeError, KeyError):
+                        # python2 - MultiIndex fails on mixed types
+                        pass
 
             # other iterable of some kind
             subarr = com._asarray_tuplesafe(data, dtype=object)

--- a/pandas/tests/test_index.py
+++ b/pandas/tests/test_index.py
@@ -553,6 +553,15 @@ class TestIndex(Base, tm.TestCase):
         # corner case
         self.assertRaises(TypeError, Index, 0)
 
+    def test_consruction_list_mixed_tuples(self):
+        # 10697
+        # if we are constructing from a mixed list of tuples, make sure that we
+        # are independent of the sorting order
+        idx1 = Index([('A',1),'B'])
+        self.assertIsInstance(idx1, Index) and self.assertNotInstance(idx1, MultiIndex)
+        idx2 = Index(['B',('A',1)])
+        self.assertIsInstance(idx2, Index) and self.assertNotInstance(idx2, MultiIndex)
+
     def test_constructor_from_series(self):
 
         expected = DatetimeIndex([Timestamp('20110101'),Timestamp('20120101'),Timestamp('20130101')])

--- a/pandas/tools/tests/test_merge.py
+++ b/pandas/tools/tests/test_merge.py
@@ -2002,6 +2002,13 @@ class TestConcatenate(tm.TestCase):
         result = df.append(df)
         assert_frame_equal(result, expected)
 
+    def test_with_mixed_tuples(self):
+        # 10697
+        # columns have mixed tuples, so handle properly
+        df1 = DataFrame({ u'A' : 'foo', (u'B',1) : 'bar' },index=range(2))
+        df2 = DataFrame({ u'B' : 'foo', (u'B',1) : 'bar' },index=range(2))
+        result = concat([df1,df2])
+
     def test_join_dups(self):
 
         # joining dups

